File: 0198-house-robber/0198-house-robber.java
class Solution {
    public int rob(int[] nums) {
        if (nums == null || nums.length == 0) {
            return 0;
        } else if (nums.length == 1) {
            return nums[0];
        }

        int prevMax = nums[0];
        int currentMax = Math.max(nums[0], nums[1]);

        for (int i = 2; i < nums.length; i++) {
            int temp = currentMax;
            currentMax = Math.max(currentMax, prevMax + nums[i]);
            prevMax = temp;
        }

        return currentMax;
    }
}

File: 0198-house-robber/0198-house-robber.js
var rob = function(nums) {
    if (nums.length === 0) return 0;
    if (nums.length === 1) return nums[0];
    
    // Initialize an array to store the maximum amount of money robbed up to the current house
    const dp = new Array(nums.length);
    
    // Base cases
    dp[0] = nums[0];
    dp[1] = Math.max(nums[0], nums[1]);
    
    // Iterate from the third house to the last one
    for (let i = 2; i < nums.length; i++) {
        // The maximum amount of money robbed up to the current house
        // is the maximum of:
        // 1. The money robbed from the current house + the money robbed up to two houses ago
        // 2. The money robbed up to the previous house
        dp[i] = Math.max(nums[i] + dp[i - 2], dp[i - 1]);
    }
    
    // Return the maximum amount of money robbed considering all houses
    return dp[nums.length - 1];
};
